### PR TITLE
added functions to create pre-proccessed csv, and to generate graphs using this csv

### PR DESCRIPTION
**Pull request by dgershko**

### CODE REVIEW

**Comment by Assaf-Alon** [comment]
# Example graph (5.5.1)
image

# Example graph (5.5.2)
image

**Review by Assaf-Alon — APPROVED** [review]
Nicely done.
We need to make sure more complex graphs are looking good as well (I'm specifically concerned about legend entries)

File: game_ai/csv_utils.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    data = data.groupby(["level", "time_ms", "rollout_mode", "heuristic_mode", "selection_mode", "loss_heuristic", "ucb_const"]).agg(
+        num_games=("win", "count"),
+        num_wins=("win", "sum")
+    )
+    data.reset_index(inplace=True)
+    data["win_rate"] = data["num_wins"] / data["num_games"]
+    data["lower_bound"], data["upper_bound"] = zip(*data.apply(lambda row: proportion_confint(row["num_wins"], row["num_games"], alpha=0.05, method='wilson') if row["num_games"] > 0 else (np.nan, np.nan), axis=1))
+    data["lower_bound"] = data["lower_bound"].apply(lambda x: round(x, 2) if not np.isnan(x) else np.nan)
+    data["upper_bound"] = data["upper_bound"].apply(lambda x: round(x, 2) if not np.isnan(x) else np.nan)
+    data = data[data["num_games"] > 200]
```
[the remark this answers is not in the document]
Comment (dgershko) [reply game_ai/csv_utils.py:88]: Generally agree, right now this is still in development and this lets me adjust values on the fly for experiments which don't have enough data points yet.

File: game_ai/csv_utils.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    data = pd.read_csv("data/all.csv")
+    NORMAL_MCTS = mcts.NORMAL_MCTS
+    MAX_NODE = mcts.MAX_NODE
+    AVG_NODE = mcts.AVG_NODE
+    PARALLEL_TREES = mcts.PARALLEL_TREES
+
+    NO_HEURISTIC = mcts.NO_HEURISTIC
+    # HEURISTIC_MCTS = mcts.HEURISTIC_MCTS
+    HEURISTIC_SELECT = mcts.HEURISTIC_SELECT
+    # HEURISTIC_EXPAND = mcts.HEURISTIC_EXPAND
```
[the remark this answers is not in the document]
Comment (dgershko) [reply game_ai/csv_utils.py:51]: I copied these dicts from the (pre-war, esoteric, somewhat arcane) utils.py file. 
I think we decided not to use these values in our experiments as they didn't result in meaningfully different outcomes. At a glance, the CSV which results from this looks correct, meaning we don't have experiments for these values in our dataset, but it's possible I just didn't notice the broken values.

File: game_ai/csv_utils.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        if self.heuristic_mode:
+            data = data[data["heuristic_mode"] == self.heuristic_mode]
+        if self.selection_mode:
+            data = data[data["selection_mode"] == self.selection_mode]
+        if self.loss_heuristic:
+            data = data[data["loss_heuristic"] == self.loss_heuristic]
+        if self.ucb_const:
+            data = data[data["ucb_const"] == self.ucb_const]
+        if self.time_ms:
+            data = data[data["time_ms"] == self.time_ms]
```
[the remark this answers is not in the document]
Comment (dgershko) [reply game_ai/csv_utils.py:116]: It shouldn't happen. However, we should still avoid this possibility, so I'll explicitly check for none.

**Comment by dgershko** [comment]
Considering generating graphs for all reasonable data comparisons:
1) compare rollout modes across levels with normal agents
2) compare selection modes improvement in one of the levels for each rollout mode (as we happen to know it does some weird stuff for parallel trees)
3) compare each heuristic and then combinations of all three for each level (9, 9+, 9++), for each of the rollout modes
4) compare select combinations of the above to demonstrate effectiveness of improvements

Note that in my experience, level 9 usually proves to be not very interesting, resulting in rapid approach to 100% regardless of params used.
Therefore, a possibility is to dump level 9 for most of these graphs, and then create side-by-side combined graphs (this lets us add two graphs next to each other in the doc. Adding two separate graphs shrinks them beyond usefulness)

**Review by AssafAlon1 — APPROVED** [review]
image

LGTM 👍

**Pull request merged**